server/controllers/firestore.py

```diff
@@ -12,59 +12,110 @@ def __init__(self, project_id=None, oauth2_credentials=None):
         )
         client = db.collection('students')
         self.metadata = client.document('metadata')
+        self.testimonials = db.collection('testimonials')
+        self.testimonials_metadata = self.testimonials.document('metadata')
         super(FirestoreDriver, self).__init__(client)
 
-    def add(self, data):
+    def _add(
+        self,
+        data,
+        *,
+        validation_fields,
+        client,
+        metadata,
+        check_unique=False,
+        check_exists=False,
+        validation_client=None
+    ):
         found_item = None
-        try:
-            found_item = self._find({'email': data['email']})
-            found_item = self._find({'student_number': data['student_number']})
-        except StopIteration:
-            pass
-
-        if found_item is not None:
-            return None
-
-        _, doc_ref = self.database_client.add(data)
-        self.metadata.update({
+        for field in validation_fields:
+            try:
+                found_item = self._find({field, data[field]}, validation_client)
+            except StopIteration:
+                pass
```

Comment:
Might be better to handle this in `_find`?  As is, it requires knowledge of the inner workings of `_find` which breaks the "functions-as-black-boxes" ideal

Reply:
Thats fair, though since this is a function class I think there's a bit more leeway. I'll change it over though